pkg/storage/chunk/objectclient/client.go

```diff
@@ -91,19 +90,22 @@ func (o *Client) getChunk(ctx context.Context, decodeContext *chunk.DecodeContex
 		key = o.keyEncoder(key)
 	}
 
-	readCloser, err := o.store.GetObject(ctx, key)
+	readCloser, size, err := o.store.GetObject(ctx, key)
 	if err != nil {
 		return chunk.Chunk{}, errors.WithStack(err)
 	}
 
 	defer readCloser.Close()
 
-	buf, err := ioutil.ReadAll(readCloser)
+	// adds bytes.MinRead to avoid allocations when the size is known.
+	// This is because ReadFrom reads bytes.MinRead by bytes.MinRead.
+	buf := bytes.NewBuffer(make([]byte, 0, size+bytes.MinRead))
```

Comment:
I'm not sure I understand the last comment. Is there some overhead to `buf.ReadFrom`?

Reply:
We add `bytes.MinRead` to the allocation to account for how readFrom works https://cs.opensource.google/go/go/+/refs/tags/go1.17.4:src/bytes/buffer.go;l=202

Readfrom will read MinRead by minRead making sure the buffer has enough for a read. If you're at the end-1 bytes then it will try to alloc end+MinRead, although it only needed 1 byte for the last read. This is to avoid the allocation and the memmove.

Reply:
Ah. Ok thanks :slightly_smiling_face: